sympy/polys/polyclasses.py

```diff
@@ -1015,29 +1185,1154 @@ def eq(f, g, strict=False):
+        return tuple(f.to_list())
+
+    def _convert(f, dom):
+        """Convert the ground domain of ``f``. """
+        if dom == QQ and f.dom == ZZ:
+            return f.from_rep(flint.fmpq_poly(f._rep), dom)
+        elif dom == ZZ and f.dom == QQ:
+            # XXX: python-flint should provide a faster way to do this.
+            return f.to_DMP_Python()._convert(dom).to_DUP_Flint()
+        else:
+            raise RuntimeError(f"DUP_Flint: Cannot convert {f.dom} to {dom}")
+
+    def _slice(f, m, n):
+        """Take a continuous subsequence of terms of ``f``. """
+        coeffs = f._rep.coeffs()[m:n]
+        return f.from_rep(f._cls(coeffs), f.dom)
+
+    def _slice_lev(f, m, n, j):
+        """Take a continuous subsequence of terms of ``f``. """
+        # Only makes sense for multivariate polynomials
+        raise NotImplementedError
+
+    def _terms(f, order=None):
+        """Returns all non-zero terms from ``f`` in lex order. """
+        if order is None or order.alias == 'lex':
+            terms = [ ((n,), c) for n, c in enumerate(f._rep.coeffs()) if c ]
+            return terms[::-1]
+        else:
+            # XXX: InverseOrder (ilex) comes here. We could handle that case
+            # efficiently by reversing the coefficients but it is not clear
+            # how to test if the order is InverseOrder.
+            #
+            # Otherwise why would the order ever be different for univariate
+            # polynomials?
+            return f.to_DMP_Python()._terms(order=order)
+
+    def _lift(f):
+        """Convert algebraic coefficients to rationals. """
+        # This is for algebraic number fields which DUP_Flint does not support
+        raise NotImplementedError
+
+    def deflate(f):
+        """Reduce degree of `f` by mapping `x_i^m` to `y_i`. """
+        # XXX: Check because otherwise this segfaults with python-flint:
+        #
+        #  >>> flint.fmpz_poly([]).deflation()
+        #  Exception (fmpz_poly_deflate). Division by zero.
+        #  Aborted (core dumped
+        #
+        if f.is_zero:
+            return (1,), f
+        g, n = f._rep.deflation()
+        return (n,), f.from_rep(g, f.dom)
+
+    def inject(f, front=False):
+        """Inject ground domain generators into ``f``. """
+        # Ground domain would need to be a poly ring
+        raise NotImplementedError
+
+    def eject(f, dom, front=False):
+        """Eject selected generators into the ground domain. """
+        # Only makes sense for multivariate polynomials
+        raise NotImplementedError
+
+    def _exclude(f):
+        """Remove useless generators from ``f``. """
+        # Only makes sense for multivariate polynomials
+        raise NotImplementedError
+
+    def _permute(f, P):
+        """Returns a polynomial in `K[x_{P(1)}, ..., x_{P(n)}]`. """
+        # Only makes sense for multivariate polynomials
+        raise NotImplementedError
```

Comment:
Why does so many things doesn't make sense for `flint` defined in the base class.
I think that they can just move to `DMP_Python` if they are defined only for them.

Reply:
I also question if it is a good idea but I did it because I thought that someone might be doing
```python
p = DMP(...)

if isinstance(p, DMP):
    ...
```
and I wanted that code to continue to work.

More generally I question whether it makes sense to try to define a generic interface over both univariate and multivariate polynomials. That is how Poly and DMP work though and I am not trying to change that here.

I could remove these `NotImplementedError` methods but I think that it is clearer to have them here to show that they are not implemented in this class.

Reply:
> `p = DMP(...)`

I think that making `DMP` not directly instanciate, and instanciating the subclass, is more explicit and predictable,
than changing the class implicitly whether `flint` is installed or not

Reply:
> Why does so many things doesn't make sense for `flint` defined in the base class.

In most cases it is because the Flint polynomials that we are using here are univariate polynomials. In Flint univariate polynomials and multivariate polynomials are different types. Currently python-flint does not expose Flint's multivariate polynomial types but work is ongoing to make that happen.

Reply:
> I think that making `DMP` not directly instanciate, and ask users to instanciate the subclass, is more explicit and predictable,
> than changing the class implicitly whether `flint` is installed or not

Almost all users of the DMP class should prefer to use the Flint version if possible but will want to fall back on the non-Flint version otherwise. All of SymPy's internal use of DMP (and DMP is mostly internal) should be using an interface that would use the Flint version when available and fallback otherwise.

 What code should anyone write to achieve that if `DMP(...)` does not do it automatically?

Reply:
I don't think that defining `__new__` is only solution though. 
We can define factory function other place, and keep reusing from that.
For instance, plotting backends used to take that approach, but eventually had to drop it due to problems that it turned out to be messing up `__init__` and `__new__` in subclasses.

Reply:
In this PR DMP basically is a factory function. It is also a type though: the common type of both DMP_Python and DUP_Flint. It is also something that works with `isinstance` which an actual function would not be able to do do.

If we change it to a function instead of a class then what does that achieve? If we make a function with a different name then what does that achieve?

You are making many different suggestions here but I don't see what actual problems you are trying to solve or how you think that any of your suggestions solves them.

Note that we will still have implicit conversions regardless of the constructor:
```python
In [1]: from sympy.polys.polyclasses import DMP

In [2]: p1 = DMP([[1, 2], [3, 4]], ZZ)

In [3]: p2 = DMP([[3, 2], [3, 4]], ZZ)

In [4]: p1
Out[4]: DMP_Python([[1, 2], [3, 4]], ZZ)

In [5]: p2
Out[5]: DMP_Python([[3, 2], [3, 4]], ZZ)

In [6]: p1.resultant(p2, 1)
Out[6]: 
(DUP_Flint([-6, -8, 0], ZZ),
 [DMP_Python([[1, 2], [3, 4]], ZZ),
  DMP_Python([[3, 2], [3, 4]], ZZ),
  DMP_Python([[-6, -8, 0]], ZZ)])

In [7]: p1
Out[7]: DMP_Python([[1, 2], [3, 4]], ZZ)

In [9]: p1.discriminant()
Out[9]: DUP_Flint([1], ZZ)
```
Is is absolutely intended that when we drop to the base case of univariate polynomials we will use the Flint version.

Reply:
> You are making many different suggestions here but I don't see what actual problems you are trying to solve or how you think that any of your suggestions solves them.

It is not the problem about whether my suggestion solves the problem,
any nasty logic can solve the problem, 
but about more right way to solve them.

I just see the problem of how the object-oriented programming in Python can be misunderstood in that regards.

Of course, you may think that DMP is just an union of two classes, which are hypothetically final (not subclassable)
and it works fine in that regards, 
but Python assumes to subclass DMP arbitrary.
And if you have 3rd subclass of DMP, it won’t be possible to use that without modifying the base class.

Reply:
Also:
```python
In [10]: DMP([ZZ(1), ZZ(2)], ZZ)
Out[10]: DUP_Flint([1, 2], ZZ)

In [11]: DMP([ZZ(1), ZZ(2)], ZZ).convert(ZZ[x])
Out[11]: DMP_Python([1, 2], ZZ[x])

In [12]: DMP([ZZ(1), ZZ(2)], ZZ).convert(ZZ[x]).convert(ZZ)
Out[12]: DUP_Flint([1, 2], ZZ)
```
The representation will convert from DUP_Flint to DMP_Python and back as needed so that we always use the fastest available representation. A user of DMP should not need to care about this: they use `DMP(...)` and they get an instance of `DMP`. It will always be the fastest available implementation of `DMP`.

The design of the methods here is that the subclass methods do not change type implicitly but the public superclass methods will to choose the best possible implementation. A user of the `DMP` class should only care that they use the public methods and the class does its best to make those as fast as possible.

Reply:
Okay, if I get that it manages the choice of its subclass by its own dynamically, then I can understand its scope.
I'm sorry that It's not possible get this idea just by reading the short manifest

Reply:
> if you have 3rd subclass of DMP, it won’t be possible to use that without modifying the base class.

I agree but that seems fine to me. We can modify the base class when needed. This class is not designed to be subclassed in downstream codebases. If others choose to subclass a class that was not intended to be subclassed then they can deal with the consequences.

Reply:
> It's not possible get this idea just by reading the short manifest

Sorry, I should have explained the design in more detail. Ideally I would put this in the docs but I don't want to do that yet because I anticipate many more changes in a relatively short time. It would be premature to write detailed docs explaining this structure right now.

For now the basic intention is just that for anyone using the `DMP` class they should use it and not worry that it is an abstract superclass with two concrete subclasses. The DMP methods should manage that implicitly and do what is best without the user of the class needing to know what is happening.